src/utils/date.js

```diff
@@ -51,3 +53,49 @@ export function getDate(value) {
   // Must be a `Date` object.
   return value.getTime();
 }
+
+/** Map from attribute names to their parsers. */
+const dateAttrParsers = {
+  'datetime': (datetime) =>
+    userAssert(parseDate(datetime), `Invalid date: ${datetime}`),
+  'end-date': (datetime) =>
+    userAssert(parseDate(datetime), `Invalid date: ${datetime}`),
+  'timeleft-ms': (timeleftMs) => Date.now() + Number(timeleftMs),
+  'timestamp-ms': Number,
```

Comment:
Nit: https://jakearchibald.com/2021/function-callback-risks/
```suggestion
  'timestamp-ms': (ms) => Number(ms),
```